### ISSUE TITLE
initialize vs connect

### ISSUE THREAD

**Issue by saneef** [stimulusjs/stimulus#75]
How do the `initialize()` and `connect()` differs in use?

**Comment by sstephenson** [comment]
Stimulus calls `initialize()` once per controller, and `connect()` every time the controller is connected to the DOM.

For example, say you have:

```html
<div id="hello" data-controller="hello"></div>
```

Stimulus calls the corresponding controller’s `initialize()` method first, followed by `connect()`.

If you then remove the element from the document:

```js
const element = document.getElementById("hello")
element.remove()
```

Stimulus calls the controller’s `disconnect()` method.

Now, if you add the element back to the document:

```js
document.body.appendChild(element)
```

Stimulus calls the controller’s `connect()` method again, but does not call `initialize()`.

**Comment by saneef** [comment]
Thanks @sstephenson for the clarification. It would be great this info see in Handbook. Pardon me if I overlooked.

**Comment by sstephenson** [comment]
Yes, I agree! I’m working on getting it into the handbook right now :)

**Comment by chemturion** [comment]
Here's the link to the section in the handbook: [Lifecycle Callbacks Explained](https://stimulusjs.org/handbook/managing-state#lifecycle-callbacks-explained).

**Comment by tleish** [comment]
Wow, I also misunderstood the difference between initialize and connect.  I only recently realized my misunderstanding  when reading [a thread on discuss.hotwire.dev](https://discuss.hotwire.dev/t/persist-stimulus-controller-when-moving-element/2596/5) where another developer had the same confusion.

The documentation states initailize() is "Invoked by Stimulus":

> once, when the controller is first instantiated

My misunderstanding was around when the controller is first initialized.  I thought this meant a controller was only initialized once when the controller class was loaded into memory (initial page load or after turbo cache invalidated).

Now I understand this to mean:

> once for each HTML element that instantiates the controller

Is this correct?  If so, the documentation could use some clarification.  Adding @sstephenson explanation and example to the documentation would be helpful, although I might reword it to:

> Stimulus calls `initialize()` once for each HTML element that instantiates the controller...

**Comment by sstephenson** [comment]
I would love to improve our explanation of initialize vs connect in the documentation, but "HTML element that instantiates the controller" is not accurate. The element does not instantiate anything. When an element with a data-controller attribute is added to the document, or when a data-controller attribute is added to an existing element, Stimulus notices and instantiates the controller in response.

**Comment by tleish** [comment]
Perhaps something like:


| Method       | Invoked by Stimulus…                                         |
| ------------ | ------------------------------------------------------------ |
| initialize() | Anytime an element with a controller is connected to the DOM *for the first time* |
| connect()    | Anytime an element with a controller is connected to the DOM |
| disconnect() | Anytime an element with a controller is disconnected from the DOM |

**Comment by javan** [comment]
In [this comment](https://discuss.hotwire.dev/t/persist-stimulus-controller-when-moving-element/2596/5) I wrote

>  You can think of `initialize()` as “first connect” if that helps.

which does seem to help?

**Comment by sstephenson** [comment]
> Anytime an element with a controller is connected to the DOM for the first time

Or maybe "Just before an element with a controller is connected to the DOM for the first time"? I don't want to introduce ambiguity around whether or not `connect()` is also called the first time a controller connects.

**Comment by tleish** [comment]
That works.  I might consider adding "each".

> Just before each element with a controller is connected to the DOM for the first time

This makes it clear you are initializing the element/controller instance, not just the controller class.

**Comment by alex-kovshovik** [comment]
The documentation still confuses me about `initialize` vs `connect`: if I use the same controller twice on a page, both `initialize` and `connect` are called twice. An example with an element being removed from DOM and then re-added makes sense, but can you help me understand the use cases of removing a DOM element and then re-adding **the same** DOM element?

In Rails Turbo, for example, when a new Turbo Frame HTML is received from the server, doesn't this create new elements? Or is there some caching algorithm that re-uses the DOM elements?

**Comment by bellef** [comment]
Possible duplicate of #215 , doc is still not crystal clear IMO

**Comment by panda-67** [comment]
### Stimulus Controller Lifecycle

| Stage          | Purpose                                   | When to Use   |
|----------------|-------------------------------------------|---------------|
| **initialize()** | Set up internal variables only, no DOM touching! | Pre-DOM |
| **connect()**    | Safe to manipulate DOM and targets         | Post-DOM |
| **disconnect()** | Cleanup (event listeners, observers, etc.) | Leaving DOM |

---
**Rule of thumb:**  
- Use `initialize()` only for setting up internal data or state — **never for DOM interaction**.  
- DOM and targets are safe to access only inside `connect()`.
- Always cleanup observers, listeners, timers, etc. inside `disconnect()`.

🔥 If you're touching the page → **use `connect()`**, not `initialize()`.
